# DTS 201: Introduction to Data Science
**Student Name:** Chifumnaya Jasper Omefe

# Week 1: Basic Descriptive Statistics

In [5]:
import statistics

# Input data
ages = [19, 22, 21, 35, 28, 24, 31, 19, 27, 40]

# Total
count = len(ages)
print(f"a. Number of responses: {count}")

# Sum of ages
sumAges = sum(ages)
print(f"b. Sum of ages: {sumAges}")

# Mean age 
meanAges = sumAges / count
print(f"c. Mean age: {meanAges:.1f}")

# Min and Max ages
minAges = min(ages)
maxAges = max(ages)
print(f"d. Min: {minAges} | Max: {maxAges}")

# Median
median_age = statistics.median(ages)
print(f"e. Median age: {median_age}")

a. Number of responses: 10
b. Sum of ages: 266
c. Mean age: 26.6
d. Min: 19 | Max: 40
e. Median age: 25.5


# Week 2: Basic Descriptive Statistics

In [6]:
import pandas as pd

# Load dataset
df = pd.read_csv("50_Startups.csv")

# Print shape
print(f"Shape: {df.shape}")
print(f"Number of rows: {df.shape[0]}")
print(f"Number of columns: {df.shape[1]}")
df.info()

Shape: (50, 5)
Number of rows: 50
Number of columns: 5
<class 'pandas.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   R&D Spend        50 non-null     float64
 1   Administration   50 non-null     float64
 2   Marketing Spend  50 non-null     float64
 3   State            50 non-null     str    
 4   Profit           50 non-null     float64
dtypes: float64(4), str(1)
memory usage: 2.1 KB


### 02: Column Data Types
R&D Spend: Continuous (Quantitative)
Administration: Continuous (Quantitative)
Marketing Spend: Continuous (Quantitative)
State: Nominal (Qualitative)
Profit: Continuous (Quantitative)

In [7]:
# Nulls per column
missing_counts = df.isnull().sum()
print(missing_counts)

R&D Spend          0
Administration     0
Marketing Spend    0
State              0
Profit             0
dtype: int64


### 03: Handling Missing Values
If there were missing values in numeric spend columns, I would impute them using the median of that column without removing observations.

In [8]:
# Single brackets -> pandas Series
series_profit = df['Profit']
print(f"Series Type: {type(series_profit)}")
print(f"Series Shape: {series_profit.shape}")

print("-" * 40)

# Double brackets -> pandas DataFrame
df_profit = df[['Profit']]
print(f"DataFrame Type: {type(df_profit)}")
print(f"DataFrame Shape: {df_profit.shape}")

Series Type: <class 'pandas.Series'>
Series Shape: (50,)
----------------------------------------
DataFrame Type: <class 'pandas.DataFrame'>
DataFrame Shape: (50, 1)


In [9]:
# Summary statistics
df.describe()

,R&D Spend,Administration,Marketing Spend,Profit
count,50.000000,50.000000,50.000000,50.000000
mean,73721.615600,121344.639600,211025.097800,112012.639200
std,45902.256482,28017.802755,122290.310726,40306.180338
min,0.000000,51283.140000,0.000000,14681.400000
25%,39936.370000,103730.875000,129300.132500,90138.902500
50%,73051.080000,122699.795000,212716.240000,107978.190000
75%,101602.800000,144842.180000,299469.085000,139765.977500
max,165349.200000,182645.560000,471784.100000,192261.830000


### 05: Summary Statistics & Outlier Check
* **Mean Values:** 
  * R&D Spend: ~\$73,721.62
  * Administration: ~\$121,344.64
  * Marketing Spend: ~\$211,025.10
  * Profit: ~\$112,012.64

**Values Looking Out of Place:**
Yes. The minimum value for **R&D Spend** and **Marketing Spend** is **0.00**. Having a startup record $0 spent on R&D or Marketing could indicate either genuine bootstrapping, missing records recorded as zero, or early-stage entry errors.

In [10]:
# Export to JSON
df.to_json("50_startups.json", orient="records", indent=4)

# Display sample of JSON format
with open("50_startups.json") as f:
    sample_json = [next(f) for _ in range(12)]
print("".join(sample_json))

[
    {
        "R&D Spend":165349.2,
        "Administration":136897.8,
        "Marketing Spend":471784.1,
        "State":"New York",
        "Profit":192261.83
    },
    {
        "R&D Spend":162597.7,
        "Administration":151377.59,
        "Marketing Spend":443898.53,



### 07: Irrelevant Features for Predicting Grades

1. **Student Matriculation / ID Number:** A nominal identifier used  for indexing; numeric magnitude has zero causal relationship with academic performance.
2. **Shoe Size:** A physical measurement with no cognitive or educational correlation to learning comprehension or study habits.
3. **Favorite Color:** A personal aesthetic preference that does not influence academic engagement, discipline, or exam mastery.

# Week 3: Data Collection Smapling

In [1]:
import pandas as pd

# Load Excel file
df_messy = pd.read_excel("50_startups_messy.xlsx")

# 1. Shape
print(f"Shape: {df_messy.shape}")
print(f"Rows: {df_messy.shape[0]}, Columns: {df_messy.shape[1]}")

# 2. Missing counts per column
print("\n--- Missing Values Count ---")
print(df_messy.isnull().sum())

# 3. Duplicate rows count
num_duplicates = df_messy.duplicated().sum()
print(f"\nDuplicate rows count: {num_duplicates}")

Shape: (50, 5)
Rows: 50, Columns: 5

--- Missing Values Count ---
R&D Spend          0
Administration     0
Marketing Spend    1
State              0
Profit             0
dtype: int64

Duplicate rows count: 0


In [5]:
# Make a copy
df_clean = df_messy.copy()

# a. Drop duplicate rows
df_clean = df_clean.drop_duplicates()

# 1. Clean all column names (removes hidden leading/trailing spaces and standardizes)
df_clean.columns = df_clean.columns.str.strip()

# Check to confirm 'State' is now present
print("Cleaned Columns:", df_clean.columns.tolist())

# b. Standardise 'State' (strip leading/trailing whitespaces, capitalize appropriately)
df_clean['State'] = df_clean['State'].astype(str).str.strip().str.title()

# Fix typos if present (e.g. 'Cal', 'Ny', 'Fl')
state_mapping = {
    'Ny': 'New York',
    'Nyc': 'New York',
    'Newyork': 'New York',
    'Cal': 'California',
    'Calif': 'California',
    'Fl': 'Florida',
    'Fla': 'Florida'
}
df_clean['State'] = df_clean['State'].replace(state_mapping)

# c. Coerce 'Marketing Spend' to numeric (turn invalid text/strings into NaN)
df_clean['Marketing Spend'] = pd.to_numeric(df_clean['Marketing Spend'], errors='coerce')

# Verify results
print("Cleaned DataFrame info:")
df_clean.info()
print("\nUnique States:", df_clean['State'].unique())

Cleaned Columns: ['R&D Spend', 'Administration', 'Marketing Spend', 'State', 'Profit']
Cleaned DataFrame info:
<class 'pandas.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   R&D Spend        50 non-null     float64
 1   Administration   50 non-null     float64
 2   Marketing Spend  47 non-null     float64
 3   State            50 non-null     str    
 4   Profit           50 non-null     float64
dtypes: float64(4), str(1)
memory usage: 2.1 KB

Unique States: <StringArray>
['New York', 'California', 'Florida']
Length: 3, dtype: str


In [6]:
print(df_messy.columns.tolist())

['R&D Spend', 'Administration', 'Marketing Spend', 'State', 'Profit']


In [7]:
# 1. Simple Random Sample (n = 20, random_state = 7)
simple_sample = df_clean.sample(n=20, random_state=7)

# 2. Stratified 20% Sample by 'State'
stratified_sample = df_clean.groupby('State', group_keys=False).apply(
    lambda x: x.sample(frac=0.20, random_state=7)
)

print(f"Simple Random Sample size: {len(simple_sample)}")
print(f"Stratified Sample size: {len(stratified_sample)}")
print("\nStratified Sample State Breakdown:\n", stratified_sample['State'].value_counts())

Simple Random Sample size: 20
Stratified Sample size: 9


KeyError: 'State'

In [8]:
# Calculate means for R&D Spend and Marketing Spend
features = ['R&D Spend', 'Marketing Spend']

pop_means = df_clean[features].mean()
simple_means = simple_sample[features].mean()
stratified_means = stratified_sample[features].mean()

# Build comparison table
comparison_df = pd.DataFrame({
    'Population Mean': pop_means,
    'Simple Random Mean (n=20)': simple_means,
    'Stratified Mean (20%)': stratified_means,
    'Diff (Simple - Pop)': simple_means - pop_means,
    'Diff (Stratified - Pop)': stratified_means - pop_means
})

comparison_df

,Population Mean,Simple Random Mean (n=20),Stratified Mean (20%),Diff (Simple - Pop),Diff (Stratified - Pop)
R&D Spend,73721.615600,76076.374500,44181.638889,2354.758900,-29539.976711
Marketing Spend,207414.790851,239540.931053,146787.405556,32126.140202,-60627.385296
